### ISSUE TITLE
Does the spec need to mention \0 in strings?

### ISSUE THREAD

**Issue by greggman** [gpuweb/gpuweb#3550]
Example:

```js
    const module = device.createShaderModule({
        code: `
            // should \0 this compile?

            @vertex fn vs(
              @builtin(vertex_index) VertexIndex : u32
            ) -> @builtin(position) vec4<f32> {
              var pos = array<vec2<f32>, 3>(
                vec2(0.0, 0.5),
                vec2(-0.5, -0.5),
                vec2(0.5, -0.5)
              );

              return vec4(pos[VertexIndex], 0.0, 1.0);
            }

            @fragment fn fs() -> @location(0) vec4<f32> {
              return vec4(1.0, 0.0, 0.0, 1.0);
            }
        `,
    });
```

Will this cause issues?

```js
  device.createBuffer({
    label: "null\0buffer",
    size: 1024,
    usage: GPUBufferUsage.UNIFORM | GPUBufferUsage.COPY_DST,
  });
```

**Comment by greggman** [comment]
Chrome fails to compile the shader above as it appears to think the shader is just `// should ` whereas Firefox succeeds in compiling the shader above.

**Comment by Kangz** [comment]
I think the spec allows zeroes and that Firefox has the correct behavior there, but we could check. We have a [dawn issue](https://bugs.chromium.org/p/dawn/issues/detail?id=1345) to track this bug already.

**Comment by ben-clayton** [comment]
https://www.w3.org/TR/WGSL/#textual-structure

> The program text must not include a null code point (U+0000).

**Comment by kdashg** [comment]
I feel like traditionally this is solved by just not having U+0000 be part of the character set, by not being recognized by any rule.
But we did forbid this, it looks like.

**Comment by greggman** [comment]
Honestly, I was more worried if there are any implementation issues. for example, if there is code like this

```
// nonsense code
void foo(const string& s) {
   const char* c = strcpy(s.c_str());
   someNativeAPIFunctionThatTakesACharPointerAndASize(c, s.size());
   ....
}
```

if you happen to forget that DOMString and USVString allow U+0000 in them.  Effectively the code above tells the native API to look at undefined memory.

For example I noticed the label issue above, if I set a label to `"foo\0bar"`, at least in Chrome, the error messages I get back prints `"foo"`, which is fine. The spec does not require error messages to use the user's label. But it's clear, somewhere, everything after the `\0` was dropped. Was it dropped safely? Probably?

Similarly, the expectation is these labels are passed to native APIs. Should the spec non-normatively mention maybe they should be sanitized and/or given a maximum length? Should to spec enforce a max length and/or test for a max length? By test, I mean if these strings have a limit in the native API then if I pass in a string beyond that limit from WebGPU something unexpected would fail.
